web/modules/custom/joinup_licence/joinup_licence.module

```diff
@@ -32,3 +44,72 @@ function joinup_licence_rdf_entity_access(EntityInterface $entity, $operation, A
   $route_match = \Drupal::routeMatch();
   return AccessResult::forbiddenIf($route_match->getRouteName() === 'entity.rdf_entity.canonical' && $route_match->getParameter('rdf_entity')->bundle() === 'spdx_licence');
 }
+
+/**
+ * Implements hook_entity_extra_field_info().
+ *
+ * - Append additional fields to the licence entity display containing data
+ *   from the SPDX licence.
+ *
+ * @see joinup_licence_rdf_entity_view()
+ */
+function joinup_licence_entity_extra_field_info(): array {
```

Comment:
I have a question on this extra fields here concerning best practices. This is to allow data from the spdx licence to be viewed in the licence page. However, only the canonical route of the spdx_licence is disallowed access which means that the spdx licence can be viewed anywhere else.
Why is it that we don't simply render the spdx_licence entity in a view mode that shows these fields and display this view mode as part of the reference field field_licence_spdx_licence?

Reply:
We need to show a number of fields from the Licence entity as well as a number of fields from the SPDX Licence entity. If I would just render the reference field to the SPDX entity using a view mode then it becomes more difficult to theme this:
* If a change to the theming is necessary then the themer needs to adapt 2 view modes instead of one.
* It would not be possible to reorder the fields freely in Field UI, for example it would not be possible to move a field from the SPDX Licence entity in between two fields of the Licence entity.
* There will be additional containers around the SPDX Licence fields which will make it necessary to write different CSS selectors for the Licence fields and the SPDX licence fields.

By embedding them like this as extra fields maximum flexibility is given to the themer to make any design possible by allowing them to position the fields freely.

![2019-04-08-173347_3520x1080_scrot](https://user-images.githubusercontent.com/442924/55737092-dd526d80-5a24-11e9-9218-327f7705d66d.png)

Reply:
The thing with the theming is that when it comes to custom templates, it is already extremely flexible. The variables array already contains both the full view of the entity and also the content where each field is contained separately in a renderable array.

In case the themer wants to highly change the way the entity looks when viewing it, he can provide an html and then simply print the fields separately. These fields can either be retrieved as raw value or as a renderable array controlled by the view mode (e.g. the title display, prefixes, suffixes etc).

The same applies for sub-entities (referenced entities) and it makes it much cleaner to simply render a view mode of the entity, because it will be a template included in the mother template.
So in this specific case, Tom for example, will theme the sub entity in its specific view mode as a separate entity because it is cleaner and easier to maintain. Then, in the parent template (the licence parent), he will only need to print the field and everything else will be done automatically. 

And last, even if it was that for some reason extra fields are required, according to what I remember from Francesco, in case of theming, they should be generated and included in the preprocess hook so that they are cached along with the theme.